packages/core-plugin-api/src/extendable-components/types.ts

```diff
@@ -0,0 +1,177 @@
+   */
+  info: ExtraProps;
+};
+
+/** @public */
+export interface ExtendableComponentRef<
+  Props extends {},
+  Context,
+  ExtraProps extends {} = {},
+> {
+  /**
+   * The 'Provider' part of an extendable component. This component should build
+   * up a context object, and forward this to the inner Component using the
+   * provided props `ComponentProvider` and `Component`.
+   */
+  Provider: ComponentType<ComponentRootProviderProps<Props, Context>>;
+
+  /**
+   * The inner 'Component' part of an extendable component. This component
+   * should render the UI (if any) by using the values of the context, and
+   * perhaps also the props.
+   */
+  Component: ComponentType<
+    ExtendableComponentProps<Props, Context> &
+      ExtendableComponentExtraProps<ExtraProps>
+  >;
+}
+
+/**
+ * The result of createExtendableComponent(). This object contains a
+ * `componentRef` and a `Component` which should both be re-exported with
+ * different names.
+ *
+ * The `componentRef` is used when _extending_ this component, and `Component`
+ * is what the users use to render the component.
+ *
+ * @public
+ */
+export interface ExtendableComponentDescriptor<
+  Props extends {},
+  Context,
+  ExtraProps extends {} = {},
+> {
+  componentRef: ExtendableComponentRef<Props, Context>;
+  Component: ComponentType<Props & { info: ExtraProps }>;
+}
+
+/** @public */
+export type RefInfoProps = Pick<RefAttributes<any>, 'ref'>;
+
+/** @public */
+export type ExtendableComponentRefConfig<Props, Context> =
+  ExtendableComponentRef<Props, Context>;
+
+/** @public */
+export type ExtendableComponentRefConfigForwardRef<Props, Context> =
+  ExtendableComponentRef<Props, Context, RefInfoProps>;
+
+/**
+ * A props interceptor function. It takes props and returns the same props or a
+ * new props object of the same type.
+ *
+ * @public
+ */
+export type ExtendableComponentPropsInterceptor<Props extends {}> = (
+  props: Props,
+) => Props;
+
+/**
+ * A props interceptor function. It takes props and returns the same props or a
+ * new props object of the same type.
+ *
+ * @public
+ */
+export type ExtendableComponentProvider<
+  Props extends {},
+  Context,
+> = ComponentType<ComponentProviderProps<Props, Context>>;
+
+/** @public */
+export type ExtendableComponentSpec<Props extends {}, Context> = {
+  /**
+   * Intercept the props, to e.g. rewrite some of them
+   */
+  interceptProps?: ExtendableComponentPropsInterceptor<Props>;
+
+  /**
+   * A provider to intercept the context between
+   */
+  Provider?: ExtendableComponentProvider<Props, Context>;
+};
+
+/**
+ * A component extension, consisting of the component ref, and the extension
+ * spec ({@link ExtendableComponentSpec}).
+ *
+ * @public
+ */
+export type ComponentExtension<Props extends {}, Context> = {
+  ref: ExtendableComponentRef<Props, Context>;
+  spec: ExtendableComponentSpec<Props, Context>;
+  plugin?: BackstagePlugin<any, any>;
```

Comment:
How do you envision this being used? eg. metadata to debug the source of plugin extensions?

Reply:
Yes and yes. Good digging.

This is currently not used, but I was thinking about e.g. error handling. Since extensions shouldn't be seen as (IMO right now) critical, but rather _nice_ improvements to existing components, it would be good if they wouldn't _fail_ the component if they threw an error e.g. Such an error could be caught and logged, and then knowing which plugin provided this extension would be useful. Especially in production builds, we could get useful messages. If we maybe also force extendable components to have id's, you could get:

`Component extension for 'link' provided by plugin 'github-actions' failed: Error: undefined is not a function`

Again, this is not implemented, but to tie a plugin to an extension was a way to _enable_ this in the future.